examples/all-clusters-app/esp32/main/Rpc.cpp

```diff
@@ -0,0 +1,329 @@
+    }
+
+    pw::Status GetIP4Address(ServerContext &, const pw_protobuf_Empty & request, chip_rpc_IP4Address & response)
+    {
+        esp_netif_ip_info_t ip_info;
+        PW_TRY(EspToPwStatus(esp_netif_get_ip_info(esp_netif_get_handle_from_ifkey("WIFI_STA_DEF"), &ip_info)));
+        snprintf(response.address, sizeof(response.address), IPSTR, IP2STR(&ip_info.ip));
+        return pw::OkStatus();
+    }
+
+    pw::Status GetIP6Address(ServerContext &, const pw_protobuf_Empty & request, chip_rpc_IP6Address & response)
+    {
+        esp_ip6_addr_t ip6;
+        PW_TRY(EspToPwStatus(esp_netif_get_ip6_linklocal(esp_netif_get_handle_from_ifkey("WIFI_STA_DEF"), &ip6)));
+        snprintf(response.address, sizeof(response.address), IPV6STR, IPV62STR(ip6));
+        return pw::OkStatus();
+    }
+
+    // NOTE: Currently this is blocking, it can be made non-blocking if needed
+    //       but would require another worker thread to handle the scanning.
+    void StartScan(ServerContext &, const chip_rpc_ScanConfig & request, ServerWriter<chip_rpc_ScanResults> & writer)
+    {
+        wifi_scan_config_t scan_config;
+        if (request.ssid_count != 0)
+        {
+            scan_config.ssid = const_cast<uint8_t *>(reinterpret_cast<const uint8_t *>(request.ssid[0].bytes));
+        }
+        if (request.bssid_count != 0)
+        {
+            scan_config.bssid = const_cast<uint8_t *>(reinterpret_cast<const uint8_t *>(request.bssid[0].bytes));
+        }
+        scan_config.channel     = request.channel;
+        scan_config.show_hidden = request.show_hidden;
+        scan_config.scan_type   = static_cast<wifi_scan_type_t>(request.active_scan);
+        if (request.active_scan)
+        {
+            scan_config.scan_time.active.min = request.scan_time_min_ms;
+            scan_config.scan_time.active.max = request.scan_time_max_ms;
+        }
+        else
+        {
+            scan_config.scan_time.passive = request.scan_time_min_ms;
+        }
+
+        auto err = esp_wifi_scan_start(&scan_config, true /* block */);
+        if (ESP_OK != err)
+        {
+            ESP_LOGI(TAG, "Error starting scan: %d", err);
+            return;
+        }
+
+        // Output scan results
+        uint16_t num_scan_records = kScanRecordsMax;
+        err                       = esp_wifi_scan_get_ap_records(&num_scan_records, scan_records);
+        if (ESP_OK != err)
+        {
+            ESP_LOGI(TAG, "Error getting scanned APs: %d", err);
+            num_scan_records = 0;
+        }
+        ESP_LOGI(TAG, "%d", num_scan_records);
+        out_scan_records.aps_count = num_scan_records;
+
+        for (size_t i = 0; i < num_scan_records; ++i)
+        {
+            memcpy(out_scan_records.aps[i].ssid.bytes, scan_records[i].ssid,
+                   std::min(sizeof(scan_records[i].ssid), sizeof(out_scan_records.aps[i].ssid.bytes)));
+            out_scan_records.aps[i].ssid.size = sizeof(out_scan_records.aps[i].ssid.bytes);
```

Comment:
Note that here the size might again be too big (bigger than the number of bytes we actually copied in).

Again, I need to understand whether this code will be shipping in production, so I know whether to review it carefully or not...

Reply:
Good catch, fixed. This is disabled by default (and for sure in production), it provides a debug interface for test automation.